query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala

```diff
@@ -111,3 +264,52 @@ object LastSampleFunction extends RangeFunction {
     }
   }
 }
+
+/**
+ * Directly obtain the last sample from chunks for much much faster performance compared to above
+ * TODO: one day we should expose higher level means of returning the actual Long value instead of converting to Double
+ */
+abstract class LastSampleChunkedFunction(var timestamp: Long = -1L,
+                                         var value: Double = Double.NaN) extends ChunkedRangeFunction {
+  override final def reset(): Unit = { timestamp = -1L; value = Double.NaN }
+  final def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = {
+    sampleToEmit.setValues(endTimestamp, value)
+  }
+  // Add each chunk and update timestamp and value such that latest sample wins
+  final def addChunks(tsCol: Int, valueCol: Int, info: ChunkSetInfo,
+                startTime: Long, endTime: Long, queryConfig: QueryConfig): Unit = {
+    val timestampVector = info.vectorPtr(tsCol)
+    val tsReader = bv.LongBinaryVector(timestampVector)
+    val endRowNum = tsReader.ceilingIndex(timestampVector, endTime)
+
+    // update timestamp only if
+    //   1) endRowNum >= 0 (timestamp within chunk)
+    //   2) timestamp is within stale window; AND
+    //   3) timestamp is greater than current timestamp (for multiple chunk scenarios)
+    if (endRowNum >= 0) {
+      val ts = tsReader(timestampVector, endRowNum)
```

Comment:
This line may return timestamp of a end-of-timeseries marker, so NaN may be returned. The actual sample returned needs to be the one prior to this value.

Reply:
Ok, will check for that.